### PR TITLE
Paper: Relation: The Missing Container (James)

### PR DESCRIPTION
**Pull request by scott-howard-james**
Relate: The Missing Container paper submission

[Edit by jaberg] http://nipy.bic.berkeley.edu:5000/download/19

### CODE REVIEW

**Comment by katyhuff** [comment]
Just FYI @scott-howard-james , your paper is failing to build with our buildbot, which struggles to know which rst file to use if there are many in the directory. Sorry about that. I can suggest that you rename your auxillary files to .txt or move them to a subdirectory, leaving only the main file in the `scott_james` directory.

```
[ 15/06 15:15 Build failed ] Relation: The Missing Container (James) by scott-howard-james 
• Updating proceedings build tools in /home/katyhuff/procbuild/procbuild/../cache/scipy_proceedings...
Already up-to-date.
• Check out paper repository...
Cloning into '/tmp/tmp5d1PFA'...
Installing proceedings build tools...
• Build the paper...
Building: scott_james
Traceback (most recent call last):
File "publisher/build_paper.py", line 179, in 
build_paper(paper_id)
File "publisher/build_paper.py", line 164, in build_paper
rst2tex(in_path, out_path)
File "publisher/build_paper.py", line 78, in rst2tex
raise RuntimeError("Found more than one input .rst--not sure which "
RuntimeError: Found more than one input .rst--not sure which one to use.
Error building paper /tmp/tmp5d1PFA/papers/scott_james. Aborting.
Error code 1
```

**Comment by scott-howard-james** [comment]
Done. (header.rst removed from folder) 

Thanks 
scott

**Comment by katyhuff** [comment]
Great! Thanks. Looks like that builds successfully (see build server : http://nipy.bic.berkeley.edu:5000/#)

**Comment by scott-howard-james** [comment]
Perfect. 

Thanks again 
scott

**Comment by scott-howard-james** [comment]
I see from website that edits are possible for a few weeks but I just wanted to make sure there was no issue/warning needed if I push edits to my proceedings fork.





Thanks


Scott





Sent with Mail Pilot

**Comment by katyhuff** [comment]
Go for it! You will soon begin to receive review comments as well. Pushing
new changes to your branch will be the right way to respond to those
comments as well.

**Comment by njsmith** [comment]
This paper:
* is 2 pages
* is mostly sentence fragment bullet points
* provides a link to the Simple English Wikipedia *instead* of describing the core concept they want to present
  * (and the Simple English Wikipedia page in fact is incomprehensible, and I had to switch to regular English Wikipedia to figure out what they were on about)
* reads like the quickstart guide to their package
  * it's just a long list of tiny toy examples
* or at least I'm assuming the pypi link thrown in at the end is to a package that
  * they wrote
  * they have been talking about in the rest of the paper
  * because it never says.

It seems to me that the main purpose of a paper like this is to provide some thoughtful prose that goes beyond basic API docs to explain to a presumed-thoughtful reader *why* something works the way it does, and what the trade-offs in that are. They conclude *"Using a relation can simplify your code and eliminate the need for repeated, ad hoc patterns when managing your working data structures"*, but I'm left entirely unconvinced of this by the text -- they don't provide a single example of any real problem that their package helped solve, or why it was superior to other mechanisms that could have been used.

Please:
* try harder
* do better.

**Comment by njsmith** [comment]
On further thought (and reading the comments above), it occurs to me that the above may be a bit harsh, because I was assuming this was actually being presented as a finished paper, and perhaps it is just that they totally blew the deadline and this is an early outline/draft. Which is still a problem given that we're in the review period now, but at least it's a different problem...

**Comment by scott-howard-james** [comment]
Thanks for your insights into Relate: The Missing Container. We've made the specific suggestions offered namely:





- Providing an in-paper definition of a mathematical relation


- Finding an easier to understand reference


- Explicitly identifying ourselves as the authors of the relation implementation referenced





Beyond these additions we're afraid our capabilities of discerning how to “do better” are limited. It's a simple concept, and as such is presented in short paper with simple, building-block examples. This intent would be obfuscated by a verbose exposition.





That said, we understand if SciPy does not deem the paper worthy of publication by virtue of either its content or its form.





Best wishes


Scott James, James Larkin


wag more. bark less.




On Jun 19, 2015, 2:55:48 AM, Nathaniel J. Smith <notifications@github.com> wrote:  

This paper:

is 2 pages
is mostly sentence fragment bullet points
provides a link to the Simple English Wikipedia instead of describing the core concept they want to present  
(and the Simple English Wikipedia page in fact is incomprehensible, and I had to switch to regular English Wikipedia to figure out what they were on about)


reads like the quickstart guide to their package  
it's just a long list of tiny toy examples


or at least I'm assuming the pypi link thrown in at the end is to a package that  
they wrote
they have been talking about in the rest of the paper
because it never says.




It seems to me that the main purpose of a paper like this is to provide some thoughtful prose that goes beyond basic API docs to explain to a presumed-thoughtful reader why something works the way it does, and what the trade-offs in that are. They conclude "Using a relation can simplify your code and eliminate the need for repeated, ad hoc patterns when managing your working data structures", but I'm left entirely unconvinced of this by the text -- they don't provide a single example of any real problem that their package helped solve, or why it was superior to other mechanisms that could have been used.


Please:

try harder
do better.

**Comment by njsmith** [comment]
From my point of view, the main problem is simply that this paper provides very little value beyond the average front-page-of-the-manual. I actually have no idea how the SciPy proceedings editors will decide what to include in the proceedings (or, indeed, whether they even will be making decisions about this at all), but to me the whole point of writing a paper or giving a talk is to go beyond what you can do in, like, docstrings. (I see I am repeating myself from the last comment, but still.)

On further investigation I guess the package in question actually has no documentation or docstrings, so maybe this is the manual. I do suggest that you'd be better off [writing a proper manual](https://read-the-docs.readthedocs.org/en/latest/getting_started.html), and... surely you have *something* more to say? Why did you even create this package? In what circumstances would it be useful? (I've never had any pressing need to keep track of fruit colors, and the standard python containers have generally worked well for me, so this isn't a rhetorical question -- I would genuinely like to understand why this package is worth my or anyone's time, and I'm sure you have reasons, but right now they're not on the page.) Why did you make the design decisions you did, e.g. the overloading of `relation[key] = value` to be a kind of append operation, instead of something that might look more conventionally pythonic like `relation.add(key, value)`? Why did you decide that for 1:1 mappings `relation[key] = value1; relation[key] = value2` should silently override instead of raising an error? What are the time and memory costs of different kinds of relations and operations on them? None of this would obfuscate the core idea of the paper.

**Comment by jaberg** [comment]
The reviewer guidelines / paper expectations are here:
https://github.com/scipy-conference/scipy_proceedings/blob/master/review_criteria.md

@jlarkin-noblis Sorry if there was misunderstsanding, but is it fair to say this isn't a fit for scipy?

**Comment by jlarkin-noblis** [comment]
Thank you Mr. Bergstra for your message.

Can you please provide a little bit more information on who you are as it relates to scipy.

>Sorry if there was misunderstsanding, but is it fair to say this . isn't a fit for scipy?

As for your question, we are a little bit confused.  Admittedly, this is our first scipy conference and we must say the information about the paper submission has been less than clear.  It was our understanding that the paper review process did not end until Tuesday and authors could update their papers based on comments they had received from the reviewers.  Scott and I have been updating our paper and were looking to upload a new version tomorrow.

Also looking over the guidelines / paper expectations, we don’t see how our paper does not meet these guidelines.  Do you have specific areas where the paper does not meet these guidelines?  Is your statement informing us that our paper does not fit for scipy and any updates we have will not change this?

Please provide additional information on what you are looking for us to do about this paper.

We will continue to move forward with our poster for the conference and look forward to your response to these questions.

Thanks


Jim J. Larkin, PMI-ACP, CSM
Lead System Analyst, Transportation Analytics
james.larkin@noblis.org

**Comment by hplgit** [comment]
I read the initial version of this paper and share the views expressed by @njsmith. Unfortunately, the new version of the paper is a minimalistic revision that does not meet the criticism of @njsmith. This is a pity, because the relation data structure is certainly useful in a number of contexts.

The fundamental problem is that the authors do not master the craft of scientific writing. They are very brief, use implicit language, illustrate points with undefined terms, use mathematical jargon that complicates the meaning of very simple things, present too trivial examples, and lack more realistic examples that can illustrate the potential of the package. Lots of typographical errors disturb the reader.

The purpose and potential of the relation data structure should be quite straightforward to explain, but we would either have to start a new paper from scratch or attack almost every sentence in the present paper and explain how it needs to be revised. Both alternatives represent a too huge undertaking for the reviewers.

@jaberg: The present version of the paper is not a fit for any scientific proceedings.

**Comment by jaberg** [comment]
Hi @jlarkin-noblis I'm one of the proceedings co-chairs.

You can get a clearer idea of what scipy papers typically look like by reading through some of the other submissions (especially ones labeled "ready") and by browsing the proceedings from previous scipy conferences.

http://conference.scipy.org/proceedings/

**Comment by jaberg** [comment]
Sorry @jlarkin-noblis , my comments should have been directed to @scott-howard-james, the owner of this pull request.

**Comment by jaberg** [comment]
@scott-howard-james and @jlarkin-noblis I've read the paper, the reviews, and had a look through the relate github repo. The issue here is a mismatch of scope. Relation is a perfectly good container, but you need to make a stronger case that a class implementing a Relation represents a significant contribution to scientific computation in Python. What scientific application should use this class? What are they doing now to solve their problem and how would they be better off using your library?

Your paper doesn't answer these questions, it is very short, and the reviewers have done their job already.  Out of respect for our reviewers and organizers, please don't upload a totally rewritten paper after the submission deadline has passed, and the review process has happened.

**Comment by jaberg** [comment]
Wow, @scott-howard-james and @jlarkin-noblis great work on the updated version!

Could you add some specific empirical timings in the section on performance?

**Comment by jaberg** [comment]
Your relation interface is a sparse matrix in some sense. Could you comment on when users should prefer your 2-dictionary data structure for implementing this interface rather than other possible data structures?

How would an implementation based on scipy's sparse matrices compare? What about networkx?

I'm skeptical that the 2-dictionary approach is competitive at scale, and part of the role of scipy is to disseminate best practices in scientific python.  You should certainly compare a few implementations to help others who use relations and dictionaries with reverse-lookup use the right tool for their jobs.

**Comment by scott-howard-james** [comment]
Can do.





Thanks


Scott





Sent with Mail Pilot




On Jun 28, 2015, 10:44:49 AM, James Bergstra <notifications@github.com> wrote:  

Wow, @scott-howard-james(https://github.com/scott-howard-james) and @jlarkin-noblis(https://github.com/jlarkin-noblis) great work on the updated version!


Could you add some specific empirical timings in the section on performance?

File: papers/scott_james/relate-scipy2015.rst [hunk 1]
```diff
@@ -0,0 +1,421 @@
...
+
+    # use the full four-letter international code ... 
+    # not the US 3-letter code 
+    airport['KATL'] = 'Hartsfield-Jackson Atlanta International'
+    airport['ATL']
+    > KeyError: 'ATL'
+
+Note that, similar to a dictionary silently overwriting a key-value
+pair, a 1:1 relation silently overwrites a value-key pair, and in this
+case removes stranded key. Also worth nothing, for cardinalities M:1 and
```
Comment (njsmith) [inline papers/scott_james/relate-scipy2015.rst]: I think you mean "noting" :-)

**Comment by njsmith** [comment]
Only had time to skim today unfortunately, but the new version is definitely much, much better. Thanks.

**Comment by scott-howard-james** [comment]
@katyhuff @jaberg @njsmith Addressed comments and pushed new version.  Thanks

**Pull request merged**